### ISSUE TITLE
It seems that there is a detail inconsistency between the code and the paper. Where does the µt that the negative sample z' needs to stay away from come from?

### ISSUE THREAD

**Issue by anwjedb** [rosewang2008/language_modeling_via_stochastic_processes#5]
Thank you for your excellent work, but I still have a question. As shown in a figure of your paper (the link is posted below), in the latent space, for a positive triplet (z0,zt,zT) and a negative sample z', the encoder makes zt close to the expected embedding µt and z' far from µt. What I understand is that µt is generated according to z0, zT, and time t, which is independent of the time of negative sample z'. In other words, when calculating the negative sample part in the loss function, it does not use the negative sample's own time step.
[https://github.com/rosewang2008/language_modeling_via_stochastic_processes/blob/main/images/encoder.png](https://github.com/rosewang2008/language_modeling_via_stochastic_processes/issues/url)
However, in the code, I found that the time t' of the negative sample itself is used (see t=self.t[idx]), even if this time is greater than T in some triplets of the same batch. These times are equivalent to using z0, zT, and a time t' greater than T to estimate µt' and stay away from it. I want to know whether this approach is different from the paper and reasonable. I look forward to and thank you for your answers.
https://github.com/rosewang2008/language_modeling_via_stochastic_processes/blob/a546add615275fc44da91751fc1b7eb5f2de4788/language_modeling_via_stochastic_processes/src/objectives/brownian_bridge.py#L105

**Comment by rosewang2008** [comment]
Thanks for your question! This is the same approach as taken in the paper. You can think of [this part](https://github.com/rosewang2008/language_modeling_via_stochastic_processes/blob/a546add615275fc44da91751fc1b7eb5f2de4788/language_modeling_via_stochastic_processes/src/objectives/brownian_bridge.py#L100-L108) of the code as estimating the normalizing constant as done in [SimCLR](https://arxiv.org/pdf/2002.05709.pdf#page=2) which samples negatives from the batch. 

Note that `idx` goes through the entire batch. You are sampling for out-of-order timesteps which either come from other documents or come out-of-order within the document. Hopefully this resolves the confusion on 
> even if this time is greater than T in some triplets of the same batch.

Otherwise happy to further clarify things.

**Comment by anwjedb** [comment]
Thank you for your answer. I think I have understood the process of
sampling negatives from the same batch. What I don't understand is that
when the time step of a sampled negative is greater than T (the time step
of the endpoint) of the original triplet, whether this time step can be
substituted into the density formula of the Brownian bridge process, and
whether it is logically out of range.

I don't know whether I don't have a deep understanding of the brown bridge
process (for example, whether I can substitute a time step later than the
endpoint), or if I missed some details in your paper. I'm sorry to bother
you again. Looking forward to your reply.

Rose E. Wang ***@***.***> 于2022年5月13日周五 01:05写道：

**Comment by rosewang2008** [comment]
Really sorry for the late response - currently catching up on the Github issues.

You're right that picking a midpoint `t` that's out of range is not well-defined (eg. variance would then be negative). Good catch! I didn't notice this impact latent recovery (empirically we still recover a bridge), but please do share your observations if something does come up!

**Comment by LHRYANG** [comment]
I have the same question. In my opinion, the time t should be fixed when sampling and computing the negative logits.....
